File: src/main/java/macro303/neptunes_pride/Connection.java
package macro303.neptunes_pride;

import com.google.gson.Gson;
import com.google.gson.GsonBuilder;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.property.SimpleObjectProperty;
import javafx.concurrent.Task;
import macro303.console.Console;
import org.jetbrains.annotations.NotNull;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.*;
import java.nio.charset.Charset;

public class Connection extends Task<Game> {
	public static final SimpleObjectProperty<Config> configProperty = new SimpleObjectProperty<>(null);
	public static final SimpleBooleanProperty connectionUnavailableProperty = new SimpleBooleanProperty(true);
	private static final String apiAddress = "http://nptriton.cqproject.net/game/";

	public static void refreshConfig() {
		configProperty.set(Config.loadConfig());
	}

	@Override
	public Game call() throws Exception {
		refreshConfig();
		HttpURLConnection connection = null;
		try {
			connection = getConnection("full");
			if (connection != null) {
				int responseCode = connection.getResponseCode();
				if (responseCode == 200) {
					String response = readAll(connection.getInputStream());
					Gson gson = new GsonBuilder()
							.serializeNulls()
							.setPrettyPrinting()
							.disableHtmlEscaping()
							.create();
					Game game = gson.fromJson(response, Game.class);
					game.format();
					return game;
				} else
					cancel();
			} else
				cancel();
		} finally {
			if (connection != null)
				connection.disconnect();
		}
		cancel();
		return null;
	}

	@Override
	public void succeeded() {
		super.succeeded();
		connectionUnavailableProperty.set(false);
		Console.displayMessage("Successful");
	}

	@Override
	public void cancelled() {
		super.cancelled();
		connectionUnavailableProperty.set(true);
		Console.displayError("Unable to Connect to API");
	}

	@Override
	public void failed() {
		super.failed();
		connectionUnavailableProperty.set(true);
		Console.displayError("Unable to Connect to API: " + getException().getLocalizedMessage());
	}

	private String getApiAddress() {
		return apiAddress + configProperty.getValue().getGameID() + "/";
	}

	private HttpURLConnection getConnection(@NotNull String address) throws IOException {
		HttpURLConnection connection;
		Console.displayMessage("API Address: " + getApiAddress() + address);
		URL url = new URL(getApiAddress() + address);
		if (configProperty.getValue().getProxyHostname() != null && configProperty.getValue().getProxyPort() != -1) {
			Proxy proxy = new Proxy(Proxy.Type.HTTP, new InetSocketAddress(configProperty.getValue().getProxyHostname(), configProperty.getValue().getProxyPort()));
			connection = (HttpURLConnection) url.openConnection(proxy);
		} else {
			connection = (HttpURLConnection) url.openConnection();
		}
		connection.setRequestMethod("GET");
		try {
			connection.connect();
			Console.displayMessage("Connected Successfully");
		} catch (ConnectException ce) {
			connection = null;
			Console.displayMessage("Connected Unsuccessfully: " + ce.getLocalizedMessage());
		}
		return connection;
	}

	@NotNull
	private String readAll(InputStream stream) throws IOException {
		BufferedReader reader = new BufferedReader(new InputStreamReader(stream, Charset.forName("UTF-8")));
		StringBuilder builder = new StringBuilder();
		int cp;
		while ((cp = reader.read()) != -1)
			builder.append((char) cp);
		return builder.toString();
	}
}

File: src/main/kotlin/macro303/neptunes_pride/Connection.kt
package macro303.neptunes_pride

import com.google.gson.GsonBuilder
import javafx.concurrent.Task
import macro303.console.Console
import macro303.neptunes_pride.game.Game
import java.io.BufferedReader
import java.io.IOException
import java.io.InputStream
import java.io.InputStreamReader
import java.net.*
import java.nio.charset.Charset

internal class Connection : Task<Game>() {

	@Throws(Exception::class)
	public override fun call(): Game? {
		refreshConfig()
		var connection: HttpURLConnection? = null
		try {
			connection = getConnection(address = "full")
			if (connection != null) {
				if (connection.responseCode == 200) {
					val response = readAll(connection.inputStream)
					val gson = GsonBuilder()
						.serializeNulls()
						.setPrettyPrinting()
						.disableHtmlEscaping()
						.create()
					return gson.fromJson(response, Game::class.java)
				} else
					cancel()
			} else
				cancel()
		} finally {
			if (connection != null)
				connection.disconnect()
		}
		cancel()
		return null
	}

	public override fun succeeded() {
		super.succeeded()
		Console.displayMessage(message = "Successful")
	}

	public override fun cancelled() {
		super.cancelled()
		Console.displayError(error = "Unable to Connect to API")
	}

	public override fun failed() {
		super.failed()
		Console.displayError(error = "Unable to Connect to API: ${exception.localizedMessage}")
	}

	private fun getApiAddress(): String {
		return "$apiAddress${config.gameID}/"
	}

	@Throws(IOException::class)
	private fun getConnection(address: String): HttpURLConnection? {
		Console.displayMessage(message = "API Address: ${getApiAddress()}$address")
		val url = URL(getApiAddress() + address)
		var connection: HttpURLConnection? = if (config.proxyHostname != null && config.proxyPort != null) {
			val proxy = Proxy(Proxy.Type.HTTP, InetSocketAddress(config.proxyHostname!!, config.proxyPort!!))
			url.openConnection(proxy) as HttpURLConnection
		} else {
			url.openConnection() as HttpURLConnection
		}
		connection!!.requestMethod = "GET"
		try {
			connection.connect()
			Console.displayMessage("Connected Successfully")
		} catch (ce: ConnectException) {
			connection = null
			Console.displayMessage("Connected Unsuccessfully: ${ce.localizedMessage}")
		}

		return connection
	}

	@Throws(IOException::class)
	private fun readAll(stream: InputStream): String {
		val reader = BufferedReader(InputStreamReader(stream, Charset.forName("UTF-8")))
		val builder = StringBuilder()
		var cp: Int = reader.read()
		while (cp != -1) {
			builder.append(cp.toChar())
			cp = reader.read()
		}
		return builder.toString()
	}

	companion object {
		lateinit var config: Config
		private const val apiAddress = "http://nptriton.cqproject.net/game/"

		private fun refreshConfig() {
			config = Config.loadConfig()
		}
	}
}